src/charm.py

```diff
@@ -10,63 +10,90 @@
 import os
 from charms.operator_libs_linux.v2 import snap  # type: ignore
 from ops.model import ActiveStatus, MaintenanceStatus
-from snap_management import SnapSpecError, SnapInstallError, SnapServiceError, install_snap, node_exporter_snap_name, opentelemetry_collector_snap_name
+from snap_management import (
+    SnapSpecError,
+    SnapInstallError,
+    SnapServiceError,
+    install_snap,
+    node_exporter_snap_name,
+    opentelemetry_collector_snap_name,
+)
+from singleton_snap import SingletonSnapManager
 
 # Log messages can be retrieved using juju debug-log
 logger = logging.getLogger(__name__)
-VALID_LOG_LEVELS = ["info", "debug", "warning", "error", "critical"]
+VALID_LOG_LEVELS = ['info', 'debug', 'warning', 'error', 'critical']
 
 SNAPS = [opentelemetry_collector_snap_name, node_exporter_snap_name]
 
+
 class OpentelemetryCollectorOperatorCharm(ops.CharmBase):
     """Charm the service."""
 
     def __init__(self, framework: ops.Framework):
         super().__init__(framework)
         self._reconcile()
 
-
     def _reconcile(self):
         self._install()
         self._remove()
         self.unit.status = ActiveStatus()
 
     def _install(self) -> None:
-        if self.hook != "install":
+        if self.hook != 'install':
             return
 
+        manager = SingletonSnapManager(self.unit.name)
         for snap_package in SNAPS:
-            self._install_snap(snap_package)
-            self._start_snap(snap_package)
+            manager.register(snap_package)
+            with manager.snap_operation(snap_package):
+                self._install_snap(snap_package)
+                self._start_snap(snap_package)
+
+            with manager.config_operation(snap_package):
+                # Merge configurations under a directory into one,
+                # and write it to the default otelcol config file.
+                # This is a placeholder for actual configuration merging logic.
+                # For example:
+                #
+                # content = merge_config()
+                # with open('etc/otelcol/config.yaml', 'w') as f:
+                #     f.write(content)
+                #     f.flush()
+                pass
 
     def _remove(self) -> None:
-        if self.hook != "remove":
+        if self.hook != 'remove':
             return
 
+        manager = SingletonSnapManager(self.unit.name)
         for snap_package in SNAPS:
-            self._remove_snap(snap_package)
+            manager.unregister(snap_package)
+            if not manager.is_used_by_other_units(snap_package):
+                with manager.snap_operation(snap_package):
```

Comment:
From first glance seems like we need to swap these two lines to avoid race condition.